etrago/cluster/gasclustering.py

```diff
@@ -0,0 +1,418 @@
+    )
+    from six import iteritems
+
+    from etrago.tools.utilities import *
+
+
+def select_dataframe(sql, conn, index_col=None):
+    """Select data from local database as pandas.DataFrame
+    Parameters
+    ----------
+    sql : str
+        SQL query to be executed.
+    index_col : str, optional
+        Column(s) to set as index(MultiIndex). The default is None.
+    Returns
+    -------
+    df : pandas.DataFrame
+        Data returned from SQL statement.
+    """
+
+    df = pd.read_sql(sql, conn, index_col=index_col)
+
+    if df.size == 0:
+        print(f"WARNING: No data returned by statement: \n {sql}")
+
+    return df
+
+
+def create_gas_busmap(etrago):
+    """
+    Create a bus map from the clustering of buses in space with a
+    weighting.
+
+    Parameters
+    ----------
+    network : pypsa.Network
+        The buses must have coordinates x,y.
+    Returns
+    -------
+    busmap : pandas.Series
+        Mapping of network.buses to k-means clusters (indexed by
+        non-negative integers).
+    """
+    # Create network_ch4 (grid nodes in order to create the busmap basis)
+    network_ch4 = Network()
+    buses_ch4 = etrago.network.buses
+    io.import_components_from_dataframe(network_ch4, buses_ch4, "Bus")
+
+    network_ch4.buses = network_ch4.buses[
+        (network_ch4.buses["carrier"] == "CH4") & (network_ch4.buses["country"] == "DE")
+    ]
+
+    # Cluster ch4 buses
+    kmean_gas_settings = etrago.args["network_clustering_kmeans"]
+
+    def weighting_for_scenario(x, save=None):
+        """ """
+        # TODO to be redefined
+        b_i = x.index
+        weight = pd.DataFrame([1] * len(b_i), index=b_i)
+
+        if save:
+            weight.to_csv(save)
+
+        return weight
+
+    # State whether to create a bus weighting and save it, create or not save
+    # it, or use a bus weighting from a csv file
+    if kmean_gas_settings["bus_weight_tocsv"] is not None:
+        weight_ch4 = weighting_for_scenario(
+            x=network_ch4.buses,
+            save="network_ch4_" + kmean_gas_settings["bus_weight_tocsv"],
+        )
+    elif kmean_gas_settings["bus_weight_fromcsv"] is not None:
+        weight_ch4 = pd.Series.from_csv(kmean_gas_settings["bus_weight_fromcsv"])
+        weight_ch4.index = weight_ch4.index.astype(str)
+    else:
+        weight_ch4 = weighting_for_scenario(x=network_ch4.buses, save=False)
+
+    weight_ch4_s = weight_ch4.squeeze()
+
+    # Creation of the busmap
+    busmap_ch4 = busmap_by_kmeans(
+        network_ch4,
+        bus_weightings=weight_ch4_s,
+        n_clusters=kmean_gas_settings["n_clusters_gas"],
+        n_init=kmean_gas_settings["n_init"],
+        max_iter=kmean_gas_settings["max_iter"],
+        tol=kmean_gas_settings["tol"],
+        n_jobs=kmean_gas_settings["n_jobs"],
+    )
+
+    # Add H2_grid buses to busmap
+    engine = db.connection(section=etrago.args["db"])
+    sql = f"""SELECT "bus_H2", "bus_CH4", scn_name FROM grid.egon_etrago_ch4_h2;"""

```

Comment:
Other tables are imported in etrago.io. If this table is needed we have to make sure that it is part of the eTraGo input database (and e.g. also publish it).

Reply:
I think the new option is better, thank you!